### PR TITLE
[codex] Add pure-text inbox promotion and continuity smoke coverage

### LINKED ISSUE #266: Pure-text memory V2 slice 4: inbox promotion and live memory smoke coverage
## Summary
Implement lightweight text-only promotion flows plus live verification so Alan's memory becomes both more human-like and continuously testable.

## Scope
- Add a pure-text candidate-memory / inbox flow for observations that should not be promoted directly into stable memory.
- Define promotion rules from inbox entries into:
  - `USER.md`
  - `MEMORY.md`
  - `topics/<slug>.md`
- Add topic-page maintenance so recurring subjects do not bloat `MEMORY.md`.
- Extend live runtime smoke coverage to include basic memory behavior, for example:
  - persist a stable identity or preference
  - end / restart session
  - answer from memory on the next session without re-discovery
  - preserve recent work continuity through handoff / summary surfaces
- Update memory docs and harness notes to match the automated live-test flow.

## Likely Files
- `crates/runtime/src/runtime/` (new inbox / promotion helpers likely needed)
- `crates/runtime/skills/memory/SKILL.md`
- `crates/alan/tests/live_runtime_smoke_test.rs`
- `crates/alan/tests/smoke_test.rs`
- `docs/skills_and_tools.md`

## Acceptance Criteria
- Observed-but-unconfirmed facts can be staged without polluting stable memory.
- Promotion into stable memory stays auditable and text-backed.
- Live smoke tests cover a minimal remember -> restart -> recall flow.
- Topic pages become the overflow surface instead of letting `MEMORY.md` grow unbounded.

## Depends On
- Foundation bootstrap surfaces.
- Episodic write path.
- Deterministic recall routing.

## Reference
- `docs/spec/pure_text_memory_contract.md`

### PR DESCRIPTION
**Pull request by realmorrisliu**
## Summary
- add a pure-text inbox/promotion helper that stages observed memory, promotes explicit confirmed facts, and keeps topic-page overflow auditable
- stage candidate memory from automatic memory flushes and capture a narrow set of direct user-stated stable facts at turn completion
- extend mock and live runtime smoke coverage for cross-session handoff continuity, and update memory/live-smoke docs to match the new behavior

## Testing
- cargo fmt --all
- git diff --check
- CARGO_TARGET_DIR=/tmp/alan-memory-pr4-rt cargo test -p alan-runtime memory_promotion --lib
- CARGO_TARGET_DIR=/tmp/alan-memory-pr4-rt cargo test -p alan-runtime test_build_memory_flush_inbox_draft_targets_memory_md --lib
- CARGO_TARGET_DIR=/tmp/alan-memory-pr4-alan cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact
- CARGO_TARGET_DIR=/tmp/alan-memory-pr4-alan cargo test -p alan --test live_runtime_smoke_test live_chatgpt_runtime_cross_session_continuity_smoke -- --ignored --exact --nocapture

Closes #266

### CODE REVIEW

File: crates/runtime/src/runtime/turn_executor.rs [hunk 1]
```diff
@@ -1617,6 +1617,14 @@ where
                 "fallback-turn-completed",
             )
             .await;
+            if let Err(err) = super::memory_promotion::capture_confirmed_turn_memory(
+                state.core_config.memory.workspace_dir.as_deref(),
+                &state.session,
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/turn_executor.rs]: Fixed in `1228f29`.

I now pass `state.core_config.memory.enabled` into `capture_confirmed_turn_memory(...)` at each turn-completion path, and `capture_confirmed_turn_memory` itself now no-ops when memory is disabled. That keeps the behavior aligned with the memory config contract even though runtime setup still resolves the workspace memory directory.

Added regression coverage in `memory_promotion.rs` for the disabled-memory path, plus reran the relevant smoke tests and clippy.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 2]
```diff
@@ -0,0 +1,803 @@
...
+    }
+
+    drafts
+}
+
+fn extract_fact_after_prefix(original: &str, lowered: &str, prefixes: &[&str]) -> Option<String> {
+    prefixes.iter().find_map(|prefix| {
+        lowered.find(prefix).and_then(|start| {
+            let start = start + prefix.len();
+            let extracted = extract_until_sentence_boundary(&original[start..]);
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `1228f29`.

I removed the lowercase-offset slicing pattern entirely. `extract_fact_after_prefix` and `extract_favorite_fact` now use Unicode-safe, char-boundary-aware case-insensitive prefix/delimiter matching, so we no longer derive byte offsets from `to_lowercase()` and then apply them back to the original string.

Also added regression tests covering Unicode text before both the `my favorite ...` and `my name is ...` extraction paths, then reran the targeted runtime tests, smoke tests, and clippy.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+    }
+    Some(&text[offset..])
+}
+
+fn char_boundary_indices(text: &str) -> impl Iterator<Item = usize> + '_ {
+    std::iter::once(0).chain(text.char_indices().skip(1).map(|(idx, _)| idx))
+}
+
+fn strip_prefix_case_insensitive_after_match<'a>(text: &'a str, prefix: &str) -> Option<&'a str> {
+    char_boundary_indices(text).find_map(|idx| strip_prefix_case_insensitive(&text[idx..], prefix))
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `46265e4`.

I tightened direct-fact extraction so it now scans only explicit declarative statement segments and matches prefixes at statement start, instead of scanning every character boundary inside the whole user turn. That prevents cases like `Can you confirm if my name is Bob?` from being silently promoted while still allowing later explicit statements such as `Intro sentence. My name is Éda.`.

I also added regression coverage for question-form prompts plus a later-sentence positive case, then reran `cargo fmt --all`, `cargo test -p alan-runtime --lib memory_promotion`, `cargo test -p alan --test smoke_test smoke_cross_session_ -- --nocapture`, and `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+    let (subject_raw, value_raw) = split_once_case_insensitive(original_tail, " is ")?;
+    let subject = extract_until_sentence_boundary(subject_raw);
+    let value = extract_until_sentence_boundary(value_raw);
+    if subject.is_empty() || value.is_empty() {
+        return None;
+    }
+    Some(format!(
+        "Favorite {}: {}",
+        subject.trim_end_matches('s'),
+        value
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `46265e4`.

`extract_favorite_fact` now preserves the subject exactly as stated instead of trimming a trailing `s`, so singular nouns like `class` no longer get mangled into `cla`. I added a regression test for `My favorite class is math.` alongside the other direct-memory extraction tests and reran the targeted tests plus clippy.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 5]
```diff
@@ -0,0 +1,956 @@
...
+        "- [{}] {} (promoted from {})",
+        promoted_stamp,
+        document.observation.trim(),
+        promoted_from
+    );
+
+    match document.frontmatter.target.as_str() {
+        MEMORY_USER_FILENAME | WORKSPACE_MEMORY_FILENAME => {
+            let existing = read_text_file_or_default(&target_path).await?;
+            if !existing.contains(document.observation.trim()) {
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `e25f831`.

I replaced the substring-based duplicate guard with exact promoted-observation matching parsed from promoted lines, so an existing `Name: Bobby` entry no longer blocks a later `Name: Bob` correction from being promoted.

I also added `promote_inbox_entry_treats_similar_facts_as_distinct_observations` as a regression test and reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 6]
```diff
@@ -0,0 +1,914 @@
...
+}
+
+fn derive_confirmed_memory_drafts(session: &Session) -> Vec<InboxEntryDraft> {
+    let Some(user_text) = session
+        .tape
+        .messages()
+        .iter()
+        .rev()
+        .find(|message| message.is_user())
+        .map(Message::text_content)
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `e25f831`.

`capture_confirmed_turn_memory(...)` now receives `state.turn_state.active_turn_message_start()` from `turn_executor`, and `derive_confirmed_memory_drafts(...)` only scans user messages within that active-turn slice. That prevents resume completions from re-promoting stale facts from earlier turns when the current completion has no new user assertion.

I added `derive_confirmed_memory_drafts_only_uses_active_turn_user_messages` as a regression test and reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 7]
```diff
@@ -0,0 +1,914 @@
...
+    if subject.is_empty() || value.is_empty() {
+        return None;
+    }
+    Some(format!("Favorite {}: {}", subject, value))
+}
+
+fn extract_until_sentence_boundary(text: &str) -> String {
+    let trimmed = text.trim();
+    let boundary = trimmed.find(['.', '\n', '!', '?']).unwrap_or(trimmed.len());
+    trimmed[..boundary]
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `e25f831`.

`extract_until_sentence_boundary(...)` now stops not only at sentence punctuation but also at clause separators and continuation phrases such as commas, semicolons, `and I`, and `but my`, so `My name is Bob and I prefer Vim.` now promotes `Name: Bob` instead of the entire compound clause.

I added `derive_confirmed_memory_drafts_stops_name_extraction_at_clause_boundary` as a regression test and reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/turn_executor.rs [hunk 8]
```diff
@@ -1617,6 +1617,16 @@ where
                 "fallback-turn-completed",
             )
             .await;
+            if let Err(err) = super::memory_promotion::capture_confirmed_turn_memory(
+                state.core_config.memory.enabled,
+                state.core_config.memory.workspace_dir.as_deref(),
+                &state.session,
+                state.turn_state.active_turn_message_start(),
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/turn_executor.rs]: Fixed in `43a4303`.

I pulled the post-turn memory work into a shared `finalize_turn_memory_best_effort(...)` helper and now invoke it from the `ToolBatchOrchestratorOutcome::EndTurn` branch as well, so non-cancelled tool-driven turn exits promote direct user facts just like the normal response-content completion paths.

While reviewing the surrounding flow, I also used the same helper for the fallback / interrupted-stream / normal completion branches so future turn-finish paths do not drift again, and I explicitly skip the helper on cancellation-driven `EndTurn` so cancelled turns still avoid memory writes.

Added `test_run_turn_promotes_direct_user_fact_when_tool_batch_ends_turn` as a regression test, then reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/turn_executor.rs [hunk 9]
```diff
@@ -1571,10 +1594,12 @@ where
                 }
                 ToolBatchOrchestratorOutcome::PauseTurn => return Ok(TurnExecutionOutcome::Paused),
                 ToolBatchOrchestratorOutcome::EndTurn { surfaces_refreshed } => {
-                    if !surfaces_refreshed {
-                        super::memory_surfaces::refresh_active_turn_memory_surfaces_best_effort(
+                    if !cancel.is_cancelled() {
+                        finalize_turn_memory_best_effort(
                             state,
+                            surfaces_refreshed,
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/turn_executor.rs:1597]: Fixed in `9303f5e`.

I added a replay-specific `finalize_replayed_tool_end_turn_best_effort(...)` helper in `agent_loop.rs` and now invoke it from both `RuntimeOpAction::ReplayApprovedToolCall` and `ReplayApprovedToolBatch` when they finish via `ToolBatchOrchestratorOutcome::EndTurn`, so approval/replay completions promote confirmed turn memory the same way as the normal tool-batch `EndTurn` path.

While reviewing the surrounding flow, I also verified those are the only replay `EndTurn` branches in `agent_loop`, and added regression coverage for both paths: `test_handle_submission_promotes_direct_user_fact_when_replayed_tool_call_ends_turn` and `test_handle_submission_promotes_direct_user_fact_when_replayed_tool_batch_ends_turn`.

Reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 10]
```diff
@@ -0,0 +1,1039 @@
...
+) -> Vec<InboxEntryDraft> {
+    let messages = active_turn_messages(session.tape.messages(), active_turn_start);
+    let mut drafts = Vec::new();
+    let mut seen_observations = HashSet::new();
+
+    for normalized in messages
+        .iter()
+        .filter(|message| message.is_user())
+        .map(Message::text_content)
+        .map(|text| normalize_whitespace(&text))
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `add75b1`.

I changed confirmed-fact parsing to preserve message line breaks instead of collapsing the whole user message through `normalize_whitespace(...)` before extraction. `derive_confirmed_memory_drafts(...)` now uses `normalize_message_for_fact_parsing(...)`, which normalizes whitespace within each line but rejoins lines with `\n`, so newline-based statement/clause boundaries still work for multiline turns.

While reviewing the surrounding code, I also verified this was the only fact-parsing path in `memory_promotion.rs` that flattened structural separators before calling `extract_fact_after_prefix(...)` / `extract_favorite_fact(...)`.

Added `derive_confirmed_memory_drafts_preserves_multiline_fact_boundaries` as a regression test for `"My name is Bob\nI prefer Vim"`, then reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 11]
```diff
@@ -0,0 +1,1062 @@
...
+    }
+    Some(format!("Favorite {}: {}", subject, value))
+}
+
+fn extract_until_sentence_boundary(text: &str) -> String {
+    let trimmed = text.trim();
+    let boundary = char_boundary_indices(trimmed)
+        .find(|&idx| {
+            let tail = &trimmed[idx..];
+            tail.starts_with(['.', '\n', '!', '?', ',', ';'])
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `8e5177f`.

I removed the unconditional comma boundary from `extract_until_sentence_boundary(...)` and replaced it with clause-specific comma delimiters, so values like `Bob, Jr.` are preserved while continuation clauses such as `Bob, and I prefer Vim` still stop at the right place.

While reviewing the surrounding punctuation handling, I also fixed the same truncation class for internal initialism periods by teaching both `declarative_statement_candidates(...)` and `extract_until_sentence_boundary(...)` to ignore embedded single-letter abbreviation periods such as `D.C.` until the real sentence-ending period.

Added regression coverage for all three cases: comma clause boundaries, commas inside name values, and initialisms inside favorite values. Reran `cargo fmt --all`, `cargo test -p alan-runtime --lib`, and `cargo clippy -p alan-runtime --all-targets -- -D warnings`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 12]
```diff
@@ -0,0 +1,1144 @@
...
+}
+
+fn extract_until_sentence_boundary(text: &str) -> String {
+    let trimmed = text.trim();
+    let boundary = char_boundary_indices(trimmed)
+        .find(|&idx| {
+            let tail = &trimmed[idx..];
+            tail.starts_with(['\n', '!', '?', ';'])
+                || (tail.starts_with('.') && !is_initialism_period(trimmed, idx))
+                || has_clause_boundary_delimiter(tail)
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `e8f414e`.

I ended up taking the broader refactor route instead of adding another period/abbreviation special-case to `extract_until_sentence_boundary(...)`. Confirmed turn-end promotion now goes through a dedicated model-mediated write-plan pass (`crates/runtime/prompts/memory_promotion.md`) and `memory_promotion.rs` validates the JSON plan before staging/promoting anything. That removes the brittle punctuation splitter from the production path, so cases like `My name is Dr. Bob.` and `The rule is use Python 3.12.` are no longer truncated by Rust-side sentence heuristics.

I also updated the pure-text memory spec / memory skill docs to make the new runtime-vs-LLM boundary explicit, and refreshed the runtime + smoke coverage around tool/replay/recall paths that now perform an extra promotion call.

Reran:
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `cargo fmt --all`
- `git diff --check`

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 13]
```diff
@@ -0,0 +1,1243 @@
...
+    session_id: &str,
+) -> Option<MemoryPromotionCandidate> {
+    let kind = normalize_memory_kind(&raw.kind)?;
+    let target = canonical_target_for_kind(kind);
+    let confidence = normalize_memory_confidence(&raw.confidence)?;
+    let observation = truncate_with_suffix(
+        raw.observation.trim(),
+        MEMORY_PROMOTION_MAX_OBSERVATION_CHARS,
+        "...",
+    );
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `0e1cf3a`.

I now normalize model-produced inline memory fields before they can reach durable promoted lines:
- `observation` is collapsed to a single line during candidate normalization and again defensively during `promote_inbox_entry(...)`, so multiline JSON-escaped text can no longer break the `- [date] ... (promoted from ...)` fact line shape.
- while reviewing the same formatting boundary, I also normalized `evidence` list items to single-line text because they are written as markdown bullets too.

I also hardened the durable promoted-line parser itself by switching `promoted_observation_from_line(...)` to split from the rightmost ` (promoted from ` suffix, so literal text containing that phrase inside the observation no longer breaks dedupe parsing.

Added regression coverage for all of that:
- `promote_inbox_entry_sanitizes_multiline_observation_before_writing`
- `parse_memory_promotion_candidates_normalizes_multiline_inline_fields`
- `promoted_observation_from_line_uses_last_promoted_from_suffix`

Reran:
- `cargo fmt --all`
- `cargo test -p alan-runtime --lib memory_promotion`
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`

File: crates/runtime/src/runtime/turn_executor.rs [hunk 14]
```diff
@@ -1641,9 +1663,11 @@ where
             )
             .await;
         } else {
-            super::memory_surfaces::refresh_turn_memory_surfaces_best_effort(
+            finalize_turn_memory_best_effort(
                 state,
+                false,
                 "turn-completed",
+                "after completed turn",
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/turn_executor.rs:1674]: Fixed in `6cdf23b`.

I moved turn-end memory promotion off the synchronous completion path instead of just reordering the event. `run_turn_with_cancel(...)` now refreshes memory surfaces inline, queues a deferred runtime action for the promotion LLM pass, then emits `TurnCompleted` and returns immediately. The runtime engine drains those deferred actions separately, and treats them as low-priority / preemptible work so a new user submission can cancel and requeue promotion instead of blocking the next turn.

While reviewing the surrounding flow, I applied the same treatment to the replay `EndTurn` paths in `agent_loop.rs`, so approval/replay completions no longer keep the turn active while waiting on the promotion model call.

Added regression coverage for both the new deferred behavior and the replay/tool EndTurn paths, then reran:
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `cargo fmt --all`
- `git diff --check`

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 15]
```diff
@@ -0,0 +1,1345 @@
...
+async fn generate_memory_promotion_response(
+    llm_client: &mut LlmClient,
+    llm_request_timeout_secs: u64,
+    request: GenerationRequest,
+) -> Result<crate::llm::GenerationResponse> {
+    if llm_request_timeout_secs == 0 {
+        return llm_client
+            .generate(request)
+            .await
+            .context("generate turn-end memory promotion plan");
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `6cdf23b`.

`generate_memory_promotion_response(...)` is now cancellation-aware even when `llm_request_timeout_secs == 0`: the zero-timeout branch uses `tokio::select!` on `cancel.cancelled()` versus `llm_client.generate(...)`, and the timed branch does the same around `tokio::time::timeout(...)`. That means a stalled promotion request can be interrupted instead of waiting forever.

With the broader refactor in the same commit, turn-end promotion now runs as a deferred runtime action, so it no longer blocks `TurnCompleted`; when a new submission arrives, the engine cancels the in-flight deferred promotion and requeues it behind the user work.

Added `run_turn_memory_promotion_job_timeout_zero_can_be_cancelled` as regression coverage, then reran runtime tests, the targeted smoke test, clippy, fmt, and `git diff --check`.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 16]
```diff
@@ -0,0 +1,1513 @@
...
+    .await?;
+    if candidates.is_empty() {
+        return Ok(());
+    }
+
+    let now = Utc::now();
+    for candidate in candidates {
+        let inbox_path = stage_inbox_entry(memory_dir, candidate.draft, now).await?;
+        if candidate.disposition == PromotionDisposition::PromoteNow {
+            promote_inbox_entry(memory_dir, &inbox_path, now).await?;
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs:328]: Fixed in `804c7ec`.

I now honor cancellation at the durable-write boundary instead of only during candidate generation. `capture_confirmed_turn_memory_for_session(...)` checks the cancellation token before entering the write loop, before each `stage_inbox_entry(...)`, and again before each `promote_inbox_entry(...)`, so a cancelled deferred promotion job stops before applying any further inbox or promoted-memory writes.

While reviewing the same path, I fixed the equivalent pre-promotion gap in the same loop, so both stage and promote side effects share the same cancellation guard.

Added `capture_confirmed_turn_memory_stops_before_writes_when_cancelled_after_generation` as a regression test, which cancels immediately after the model returns a `promote_now` write plan and verifies that neither inbox entries nor `USER.md` are written.

Reran:
- `cargo fmt --all`
- `cargo test -p alan-runtime --lib memory_promotion`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `git diff --check`

File: crates/runtime/src/runtime/engine.rs [hunk 17]
```diff
@@ -1052,131 +1068,184 @@ pub fn spawn_with_llm_client_and_tools(
...
+                            }
+                            incoming = sub_rx.recv(), if !submissions_closed => {
+                                match incoming {
+                                    Some(incoming) => {
+                                        if matches!(incoming.op, alan_protocol::Op::Interrupt) {
+                                            cancel.cancel();
+                                        } else {
+                                            requeue_action = true;
+                                            cancel.cancel();
+                                            queues.push_outer_submission(incoming);
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/engine.rs:1256]: Fixed in `ce8d065`.

I changed the outer runtime queue semantics so submissions are no longer appended behind deferred work. `RuntimeSubmissionQueues::push_outer_submission(...)` now inserts each submission ahead of the first queued deferred item, which means a foreground submission preempts not only the currently running deferred action but also any older deferred actions already waiting in `outer_queue`.

While reviewing the same queue boundary, I found the identical priority inversion in `requeue_leftover_inband_submissions(...)`: buffered in-turn submissions were also being extended onto the back of the queue and could end up behind deferred actions. I moved that path onto the same insertion helper so both direct incoming submissions and requeued leftovers stay ahead of deferred runtime work.

Added regression coverage for both cases:
- `test_push_outer_submission_inserts_before_existing_deferred_actions`
- `test_requeue_active_turn_leftovers_inserts_before_existing_deferred_actions`

Reran:
- `cargo fmt --all`
- `cargo test -p alan-runtime --lib engine`
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `git diff --check`

[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/engine.rs:1256]: Fixed in `d591bed`.

I changed runtime shutdown to drain deferred memory-promotion work instead of cancelling and dropping it on exit:
- when shutdown is requested, the outer loop now keeps pulling deferred items from the queue until none remain, rather than breaking immediately
- the active deferred-action path no longer cancels the current promotion job on shutdown, so the in-flight write-plan can finish cleanly
- while reviewing the same boundary, I also closed the adjacent gap where already-queued deferred actions after `TurnCompleted` were skipped by the old top-of-loop shutdown break

Added `test_runtime_shutdown_drains_deferred_memory_promotion_actions`, which reproduces the `TurnCompleted` -> immediate shutdown case and asserts the promoted fact is still written to `USER.md`.

Reran:
- `cargo fmt --all`
- `cargo test -p alan-runtime --lib engine`
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `git diff --check`

File: crates/runtime/src/runtime/engine.rs [hunk 18]
```diff
@@ -1043,146 +1077,192 @@ pub fn spawn_with_llm_client_and_tools(
...
+                    loop {
+                        tokio::select! {
+                            result = &mut action_fut => {
+                                drop(action_fut);
+                                if let Err(err) = result {
+                                    error!(error = %err, "Error handling deferred runtime action");
+                                }
+                                if requeue_action {
+                                    queues.push_outer_deferred(action_for_requeue);
+                                }
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/engine.rs]: Fixed in `32f080f`.

I split deferred-action completion into explicit `DeferredRuntimeActionExit::{Completed, Cancelled}` states and changed the engine to requeue only on an actual cancelled exit, not merely on a preemption attempt. That means a promotion job that wins the race and finishes successfully is no longer enqueued a second time for duplicate staging / extra LLM work.

While reviewing the surrounding runtime queue code, I also confirmed this was the only unconditional deferred requeue site in the runtime loop.

Added coverage for the new guard (`test_should_requeue_deferred_action_only_after_cancelled_exit`) and reran:
- `cargo fmt --all`
- `cargo test -p alan-runtime --lib engine`
- `cargo test -p alan-runtime --lib agent_loop`
- `cargo test -p alan-runtime --lib turn_executor`
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `git diff --check`

File: crates/runtime/src/runtime/engine.rs [hunk 19]
```diff
@@ -1043,146 +1077,192 @@ pub fn spawn_with_llm_client_and_tools(
...
+                                    }
+                                    None => {
+                                        submissions_closed = true;
+                                    }
+                                }
+                            }
+                            _ = shutdown_rx.recv() => {
+                                shutdown_requested = true;
+                                submissions_closed = true;
+                            }
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/engine.rs:1267]: I’m not changing this branch in this PR.

The current behavior here is intentional after `d591bed`: graceful shutdown now prefers draining the in-flight deferred memory-promotion job so the last turn’s writes are not dropped in the common `TurnCompleted -> shutdown` path. Cancelling immediately in this branch would reintroduce that normal-case data loss.

The bounded graceful deadline already lives one layer up in `RuntimeController::shutdown()`: it waits up to 10 seconds, then aborts the runtime if the deferred action is still hung. That means shutdown is already bounded without making the deferred branch itself eagerly cancel durable-memory work.

I agree that a future persisted/replayable deferred-job model could justify a different tradeoff here, but with the current semantics an immediate cancel would be a regression, so I’m leaving this unchanged.

File: crates/runtime/src/runtime/memory_promotion.rs [hunk 20]
```diff
@@ -0,0 +1,1607 @@
...
+
+fn extract_json_object(raw: &str) -> Option<&str> {
+    let trimmed = raw.trim();
+    let start = trimmed.find('{')?;
+    let end = trimmed.rfind('}')?;
+    (start <= end).then_some(&trimmed[start..=end])
+}
+
+fn is_topic_target(target: &str) -> bool {
+    target.starts_with(&format!("{MEMORY_TOPICS_DIRNAME}/")) && target.ends_with(".md")
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/src/runtime/memory_promotion.rs]: Fixed in `22c575b`.

I tightened topic-target validation so promotion now accepts only `topics/<single-slug>.md` paths:
- `topic_slug_from_target(...)` now parses the target as path components and rejects traversal (`..`), absolute/rooted paths, and any multi-segment topic suffix
- `is_topic_target(...)` delegates to that validated parser instead of doing a raw prefix/suffix string check
- `resolve_target_path(...)` no longer does `memory_dir.join(raw_target)` for topics; it rebuilds the path from the validated slug under `topics/`, so a crafted inbox entry cannot escape `.alan/memory`

While reviewing the same boundary, I also checked the rest of `memory_promotion.rs` for raw `join(target)` writes and confirmed this was the only target-path write site.

Added `promote_inbox_entry_rejects_topic_target_path_traversal` as regression coverage, then reran:
- `cargo fmt --all`
- `cargo test -p alan-runtime --lib memory_promotion`
- `cargo test -p alan-runtime --lib`
- `cargo test -p alan --test smoke_test smoke_cross_session_handoff_continuity_is_recalled -- --exact`
- `cargo clippy -p alan-runtime -p alan --all-targets -- -D warnings`
- `git diff --check`
